Fix edge cases of keyboard navigations around `anchor` elements with `TabsToLink`

Ahmad-S792 opened the pull request:
#### e39892e12f8d3c3a544327c87c9904401e15ef9f
<pre>
Fix edge cases of keyboard navigations around `anchor` elements with `TabsToLink`

<a href="https://bugs.webkit.org/show_bug.cgi?id=270713">https://bugs.webkit.org/show_bug.cgi?id=270713</a>

Reviewed by Ryosuke Niwa.

This patch is to fix edge cases in keyboard focus navigation around
`anchor` elements:

Partial Merge: <a href="https://chromium.googlesource.com/chromium/blink/+/53f8eba14b6362795bae47c02236ca93303264a6">https://chromium.googlesource.com/chromium/blink/+/53f8eba14b6362795bae47c02236ca93303264a6</a>

By merging above, WebKit is trying to fix case around `empty anchor`
while `TabsToLinks` enabled. The test case was failing currently in
WebKit, so this fixes it.

Merge: <a href="https://chromium.googlesource.com/chromium/blink/+/7ef69ef071ba2b49ece747b93518b9a1218f9c45">https://chromium.googlesource.com/chromium/blink/+/7ef69ef071ba2b49ece747b93518b9a1218f9c45</a>

In this merge, WebKit is aligning to allow keyboard navigation to anchors
with `TabsToLinks` being disabled if the anchors have explicit tabindex.

Beside following, this patch still retains `early return` if it is not focusable.

&gt; Changes:
* Source/WebCore/html/HTMLAnchorElement.cpp:
(HTMLAnchorElement::isKeyboardFocusable):

&gt; Test Cases &amp; Expectations:
* LayoutTests/fast/events/anchor-empty-focus-expected.txt:
* LayoutTests/fast/events/anchor-empty-focus.html:
* LayoutTests/fast/events/frame-tab-focus-expected.txt:
* LayoutTests/fast/events/frame-tab-focus.html:
* LayoutTests/fast/events/tab-focus-anchor-expected.txt:
* LayoutTests/fast/events/tab-focus-anchor-tab-to-links-expected.txt:
* LayoutTests/fast/events/tab-focus-anchor-tab-to-links.html:
* LayoutTests/fast/events/tab-focus-anchor.html:
* LayoutTests/fast/forms/focus-on-control-with-zero-size.html:
* LayoutTests/fast/forms/focus-on-control-with-zero-size-expected.txt:

&gt; Platform Specific Test Expectation:
* LayoutTests/platform/ios/TestExpectations: `iOS` don&apos;t have &apos;EventSender` keydown support

Canonical link: <a href="https://commits.webkit.org/275976@main">https://commits.webkit.org/275976@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bd80b15a1ddf02f95d3cee5de5821abd1369d1df

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43389 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22419 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45799 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46024 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39515 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/45693 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26181 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19837 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35870 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43963 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19439 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37365 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16846 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17020 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38426 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1445 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/39623 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38741 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47568 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18317 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15053 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42648 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19840 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/37660 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41309 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9666 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20019 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19471 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/html/HTMLAnchorElement.cpp:
```diff
@@ -115,47 +116,18 @@ bool HTMLAnchorElement::isInteractiveContent() const
     return isLink();
 }
 
-static bool hasNonEmptyBox(RenderBoxModelObject* renderer)
```
I don't think we should be removing this check.

Ahmad-S792 replied:
It resolves issue around `https://bugs.webkit.org/show_bug.cgi?id=31496#c7` and taking that bug as example (JSFiddle) - https://jsfiddle.net/b058spv4/, in local build, using tab, now I am able to focus both `float` and `non-float` while in Safari 17.4, I am not.

Ahmad-S792 replied:
https://github.com/WebKit/WebKit/assets/52317531/1caa6cc2-9866-4737-b339-2b23d3c0a1fb

rniwa replied:
I understand that but that's a separate bug. We need to address it separately instead of tucking it into this PR which is about making anchor focusable when tabindex is set.

Ahmad-S792 replied:
We don't have this concept in `SVGAElement` - just for my learning - any idea why? https://searchfox.org/wubkat/rev/e80ab9289841484d684b81a0ae84ad3cbd793b05/Source/WebCore/svg/SVGAElement.cpp#181

Ahmad-S792 replied:
@rniwa - Thanks for your guidance on Slack, I retained `hasNonEmptyBox` for time being and rebased this patched and tested all new tests to ensure that we don't regress anything and just pushed new commit. After this lands, I will remove `hasNonEmptyBox` with test case (will work on it). So we can align with all other browsers and have another issue down. Thanks again for your help as always.

rniwa reviewed (changes requested):
I don't think we should get rid of hasNonEmptyBox check.

Commit:
Fix edge cases of keyboard navigations around `anchor` elements with `TabsToLink`

https://bugs.webkit.org/show_bug.cgi?id=270713

Reviewed by Ryosuke Niwa.

This patch is to fix edge cases in keyboard focus navigation around
`anchor` elements:

Partial Merge: https://chromium.googlesource.com/chromium/blink/+/53f8eba14b6362795bae47c02236ca93303264a6

By merging above, WebKit is trying to fix case around `empty anchor`
while `TabsToLinks` enabled. The test case was failing currently in
WebKit, so this fixes it.

Merge: https://chromium.googlesource.com/chromium/blink/+/7ef69ef071ba2b49ece747b93518b9a1218f9c45

In this merge, WebKit is aligning to allow keyboard navigation to anchors
with `TabsToLinks` being disabled if the anchors have explicit tabindex.

Beside following, this patch still retains `early return` if it is not focusable.

> Changes:
* Source/WebCore/html/HTMLAnchorElement.cpp:
(HTMLAnchorElement::isKeyboardFocusable):

> Test Cases & Expectations:
* LayoutTests/fast/events/anchor-empty-focus-expected.txt:
* LayoutTests/fast/events/anchor-empty-focus.html:
* LayoutTests/fast/events/frame-tab-focus-expected.txt:
* LayoutTests/fast/events/frame-tab-focus.html:
* LayoutTests/fast/events/tab-focus-anchor-expected.txt:
* LayoutTests/fast/events/tab-focus-anchor-tab-to-links-expected.txt:
* LayoutTests/fast/events/tab-focus-anchor-tab-to-links.html:
* LayoutTests/fast/events/tab-focus-anchor.html:
* LayoutTests/fast/forms/focus-on-control-with-zero-size.html:
* LayoutTests/fast/forms/focus-on-control-with-zero-size-expected.txt:

> Platform Specific Test Expectation:
* LayoutTests/platform/ios/TestExpectations: `iOS` don't have 'EventSender` keydown support

Canonical link: https://commits.webkit.org/275976@main
